### PR TITLE
feat: added dependency support to k8s

### PR DESCRIPTION
**Pull request by AvineshTripathi**
Because we currently fetch all the objects separately and without any connection, some relationships are difficult to generate and would result in redundancy. In other cases, we also need both resources to be present in order to validate the relations.

To avoid this, we should have something like a stitching module that would asynchronously join the node based on a predefined condition. Open to discussion

Solution proposed:
Based on current architecture, the relationship field is sent by the resource itself. However, sometimes comparison between two resources are required to find a relationship. The solution we are using for this PR (which may change if any better approach to this problem comes) is that if a resource requires another resource to find the relationship, it will do so by making an api call and fetching all the resources of that particular service where the relationship is to be found. However, the result is cached and stored in memory for some time, so if another resource requires this resource list can use it.

### CODE REVIEW

**Comment by Azanul** [comment]
We can create a routine which will start when all the resources(k8s) from the wait group are completed

File: internal/config/load.go [hunk 1]
```diff
@@ -34,6 +35,11 @@ import (
 	"k8s.io/client-go/tools/clientcmd"
 )
 
+const (
+	CACHE_DURATION   = 3
+	CLEANUP_DURATION = 4
+)
```
Comment (Azanul) [inline internal/config/load.go]: constants declared twice

File: internal/config/load.go [hunk 2]
```diff
@@ -243,6 +249,7 @@ func Load(configPath string, telemetry bool, analytics utils.Analytics) (*models
 
 			clients = append(clients, providers.ProviderClient{
 				K8sClient: &client,
+				Cache:     cache.New(CACHE_DURATION, CLEANUP_DURATION),
```
Comment (Azanul) [inline internal/config/load.go]: nit: keep consistent code style

File: providers/k8s/k8s.go [hunk 3]
```diff
@@ -37,7 +37,7 @@ func FetchResources(ctx context.Context, client providers.ProviderClient, db *bu
 				log.Printf("[%s][K8s] %s", client.Name, err)
 			} else {
 				for _, resource := range resources {
-					_, err = db.NewInsert().Model(&resource).On("CONFLICT (resource_id) DO UPDATE").Set("cost = EXCLUDED.cost").Exec(context.Background())
+					_, err = db.NewInsert().Model(&resource).On("CONFLICT (resource_id) DO UPDATE").Set("cost = EXCLUDED.cost, relations=EXCLUDED.relations").Exec(context.Background())
```
Comment (Azanul) [inline providers/k8s/k8s.go:40]: explain

**Review by Azanul — CHANGES_REQUESTED** [review]
1. Why are we only using cache for services?
2. address lint issue
3. Let's maintain a clear document to keep track of what changes we're making and how it'll reflect on the dependency graph

File: providers/k8s/k8s.go (hunk 3 above)
Comment (AvineshTripathi) [reply providers/k8s/k8s.go:40]: if the resource or the relation was deleted, in the next cycle it will not be updated if this is not added

**Comment by AvineshTripathi** [comment]
> 1. Why are we only using cache for services?

Since we have to get list of all the services to find the relation between pod and services, a new call to the service list has to be done. However we already have a service fetcher and since services might be used for other relations, i felt it would be better if we cache it for some time to avoid multiple calls.

This I feel can be used across the project  to find relations where presence of other resource is mandatory.
> 2. address lint issue

Will do 
> 3. Let's maintain a clear document to keep track of what changes we're making and how it'll reflect on the dependency graph

where should we document

File: providers/k8s/k8s.go (hunk 3 above)
Comment (Azanul) [reply providers/k8s/k8s.go:40]: `cost = EXCLUDED.cost` means that if a conflict occurs, the cost column of the existing row will be updated with the value from the new row being inserted. Same goes for relations

**Comment by Azanul** [comment]
> Since we have to get list of all the services to find the relation between pod and services, a new call to the service list has to be done. However we already have a service fetcher and since services might be used for other relations, i felt it would be better if we cache it for some time to avoid multiple calls.

Is the case different for nodes, pods, etc.?

> where should we document

Let's create a documentation page for dependency graph that'll help users to navigate the graph easily. Like a user guide

File: providers/k8s/k8s.go (hunk 3 above)
Comment (AvineshTripathi) [reply providers/k8s/k8s.go:40]: thats correct and we want to ensure this field gets updated

**Comment by AvineshTripathi** [comment]
> > Since we have to get list of all the services to find the relation between pod and services, a new call to the service list has to be done. However we already have a service fetcher and since services might be used for other relations, i felt it would be better if we cache it for some time to avoid multiple calls.
> 
> Is the case different for nodes, pods, etc.?
> 


Didnt get you here

**Comment by Azanul** [comment]
> > > Since we have to get list of all the services to find the relation between pod and services, a new call to the service list has to be done. However we already have a service fetcher and since services might be used for other relations, i felt it would be better if we cache it for some time to avoid multiple calls.
> > 
> > 
> > Is the case different for nodes, pods, etc.?
> 
> Didnt get you here

What's stopping us from caching nodes and pods too?

**Comment by AvineshTripathi** [comment]
In case of nodes and pod we do get fields to get the relation in there as comparison is not required while in case of service its not the case. Let me know if that answers your question

**Comment by Azanul** [comment]
> In case of nodes and pod we do get fields to get the relation in there as comparison is not required while in case of service its not the case. Let me know if that answers your question

It does, thanks

**Review by Azanul — APPROVED** [review]

**Pull request merged**